lib/localenv/detect.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package localenv
+
+import (
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/ops"
+
+	"github.com/gravitational/trace"
+)
+
+// DetectCluster attempts to detect whether there is a deployed cluster.
+//
+// The behavior is as follows:
+//
+//   - If the cluster is deployed/healthy, returns nil.
+//   - If the cluster is not deployed and the node's clean, returns NotFound.
+//   - If partial state is detected, returns a specific error.
+func DetectCluster(env *LocalEnvironment) error {
+	// If the local cluster checks succeed, we have a cluster.
+	clusterErr := checkCluster(env)
+	if clusterErr == nil {
+		return nil
+	}
+	// Otherwise see if there is a partial state by looking at local packages.
+	err := checkLocalPackages(env)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return trace.NotFound("no cluster detected")
+		}
+		return trace.Wrap(err)
```

Comment:
This is supposed to show the below message as well because it's the same partial state - i.e. considering failure to query cluster **or** query packages to be equal in this context.

Reply:
Yeah, we could return an aggregate here for sure.